src/credentials.rs

```diff
@@ -31,55 +31,110 @@ pub struct Credentials {
     pub registry_tokens: HashMap<RegistryUri, String>,
 }
 
-impl Credentials {
-    fn location() -> eyre::Result<PathBuf> {
-        let home = home::home_dir().wrap_err("Failed to locate home directory")?;
+/// Error produced when locating the credentials file
+#[derive(Error, Debug)]
+pub enum LocateError {
```

Comment:
broad comment on everything. I do not like this PR. You've made a _tonne_ of implementation details public, which dramatically increases the surface area of this project.

I strongly, strongly, strongly recommend not adding specific error types for things you do not expect the user to be able to handle.

Why? Because each thiserror error is part of your public API, meaning that when you 1.0, adding a new type of error is a breaking API change.

So e.g. for reading credentials...

```
- Locate
- Io
- Toml
```

what is an API caller of this thing supposed to be doing if a 'locate' error comes back? Why are they supposed to differentiate between 'Io' and 'Toml'? But if you add a new mechanism to e.g. indicate that it's in a deprecated format, you've just made a breaking API change for the person who wants to process the fact that the credentials are missing. But what are they supposed to do about that?

Reply:
@j-baker general purpose behind this mr is to increase error messages to users! Thats why we moved away from eyre to thiserror.

We should unify the error enums where possible imo. so there should be one `buffrs::Error` wrapper which internally holds our typed errors, which keeps the error easy for users and enables good diagnostics at the same time

Reply:
My proposal for this would be to make the error types internal only and returns a `buffrs::Error` type everywhere, i do think this level is good for diagnostics but at the moment it looks a bit cumbersome for users + for maintainers

Reply:
Of course we did, that is what we are currently using! The error messages which are displayed to users turn out to be not really helpful sadly, thats why we considered transitioning to a combination of miette and thiserror

Reply:
Having an "internal" error kind / error enum is pretty common tho. The io module of the std lib does this and offers both a `std::io::Error` + detailed diagnostics on the actual thing that went wrong.

Reply:
Should probably have kept this one in draft state, as it really is still a draft! I figured I'd need significant changes.

> My proposal for this would be to make the error types internal only and returns a buffrs::Error type everywhere

+1, that makes a lot of sense, I think this is a pattern many libraries use (like `reqwest`, though I have some pet peeves with some decisions they made).

> Having an "internal" error kind / error enum is pretty common tho. The io module of the std lib does this and offers both a std::io::Error + detailed diagnostics on the actual thing that went wrong.

Also +1 on this, it's very useful to be able to access the underlying category of error, in case you want to customise error messages, write tests more robustly, etc. The end user doesn't really care about the typed error, but again this is for the sake of the library users that may compose Buffrs into other tools.

Overall, I tried to keep a balance between informative errors and keeping internal details out. Enough to distinguish a TOML syntax error from a file permission error, for example, because those two problems have very distinct solutions. That said, `thiserror` by nature does encourage more granular errors so happy to merge variants where it makes sense, too. Since I was doing this en masse I more or less created a distinct error variant for each use of `?` throughout the code, except where it was obvious it was the same error case in the same function.

In fact, I'm curious what error classes you two think would be useful to expose publicly.